# Домашнее задание 2 - предобработка признаков, pandas


### О задании

Практическое задание 2 посвящено изучению основных библиотек для анализа данных, а также линейных моделей и методов их обучения. Вы научитесь:
 * применять библиотеки NumPy и Pandas для осуществления желаемых преобразований;
 * подготавливать данные для обучения линейных моделей;
 * обучать линейную, Lasso и Ridge-регрессии при помощи модуля scikit-learn;
 * реализовывать обычный и стохастический градиентные спуски;
 * обучать линейную регрессию для произвольного функционала качества.
 

### Оценивание и штрафы

Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов. Кроме того, некоторые из заданий являются опциональными (необязательными), однако за их выполнение можно получить дополнительные баллы, которые позднее будут учитываться при проставлении оценок автоматом по курсу.

Сдавать задание после указанного срока сдачи нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов. Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце Вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник). 

Неэффективная реализация кода может негативно отразиться на оценке.


### Формат сдачи
Для сдачи задания получившийся файл \*.ipynb с решением необходимо выложить в свой репозиторий github.

## Библиотеки для анализа данных

### NumPy

Во всех заданиях данного раздела запрещено использовать циклы  и list comprehensions. Под вектором и матрицей в данных заданиях понимается одномерный и двумерный numpy.array соответственно.

In [3]:
import numpy as np

**1. (0.2 балла)** Реализуйте функцию, возвращающую максимальный элемент в векторе x среди элементов, перед которыми стоит нулевой. Для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответом является 5. Если нулевых элементов нет, функция должна возвращать None.


In [7]:
def max_element(arr):
    mask = np.concatenate(([False], arr[:-1] == 0))
    if not np.any(mask):
        return None
    return np.max(arr[mask])

x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
print(max_element(x))

y = np.array([6, 2, 3, 4])
print(max_element(y))

5
None


**2. (0.2 балла)** Реализуйте функцию, принимающую на вход матрицу и некоторое число и возвращающую ближайший к числу элемент матрицы. Например: для X = np.arange(0,10).reshape((2, 5)) и v = 3.6 ответом будет 4.

In [8]:
def nearest_value(X, v):
    flattened_matrix = X.flatten()
    index = np.abs(flattened_matrix - v).argmin()
    return flattened_matrix[index]
nearest_value(np.arange(0,10).reshape((2, 5)), 3.6)

4

**3. (0.2 балла)** Реализуйте функцию scale(X), которая принимает на вход матрицу и масштабирует каждый ее столбец (вычитает выборочное среднее и делит на стандартное отклонение). Убедитесь, что в функции не будет происходить деления на ноль. Протестируйте на случайной матрице (для её генерации можно использовать, например, функцию [numpy.random.randint](http://docs.scipy.org/doc/numpy/reference/generated/numpy.random.randint.html)).

In [12]:
def scale(X):
    mean = np.mean(X, axis=0)
    std = np.std(X, axis=0)
    std[std == 0] = 1
    return (X - mean) / std

Xx = np.random.rand(5, 3)
scaled_X = scale(Xx)

print("Матрица X:\n", Xx)
print("Матрица X после масштабирования:\n", scaled_X)

Матрица X:
 [[0.18833825 0.69106704 0.0621653 ]
 [0.79705205 0.85669064 0.53534371]
 [0.54015187 0.11223724 0.88457278]
 [0.58283604 0.60739012 0.62751454]
 [0.82744292 0.91049097 0.19249167]]
Матрица X после масштабирования:
 [[-1.73870674  0.19564936 -1.33669993]
 [ 0.91501905  0.77959388  0.25148312]
 [-0.20495329 -1.84514961  1.42364076]
 [-0.01886899 -0.09937306  0.56084664]
 [ 1.04750996  0.96927943 -0.89927059]]


**4. (0.2 балла)** Реализуйте функцию, которая для заданной матрицы находит:
 - определитель
 - след
 - наименьший и наибольший элементы
 - норму Фробениуса
 - собственные числа
 - обратную матрицу

Для тестирования сгенерируйте матрицу с элементами из нормального распределения $\mathcal{N}$(10,1)

In [13]:
def get_stats(X):
    # Находим определитель матрицы
    det = np.linalg.det(X)
    
    # Находим след матрицы
    trace = np.trace(X)
    
    # Находим наименьший и наибольший элементы матрицы
    min_element = np.min(X)
    max_element = np.max(X)
    
    # Находим норму матрицы
    frobenius_norm = np.linalg.norm(X)
    
    # Находим собственные числа матрицы
    eigenvalues = np.linalg.eigvals(X)
    
    # Находим обратную матрицу
    inverse_matrix = np.linalg.inv(X)
    
    return det, trace, min_element, max_element, frobenius_norm, eigenvalues, inverse_matrix

# Генерируем матрицу с элементами из нормального распределения N(10,1)
X = np.random.normal(10, 1, (3, 3))

det, trace, min_element, max_element, frobenius_norm, eigenvalues, inverse_matrix = get_stats(X)

print("Определитель матрицы:", det)
print("След матрицы:", trace)
print("Наименьший элемент матрицы:", min_element)
print("Наибольший элемент матрицы:", max_element)
print("Норма Фробениуса матрицы:", frobenius_norm)
print("Собственные числа матрицы:", eigenvalues)
print("Обратная матрица:\n", inverse_matrix)

Определитель матрицы: -24.58102488991184
След матрицы: 31.752586791877384
Наименьший элемент матрицы: 8.880995240673428
Наибольший элемент матрицы: 12.949189668645532
Норма Фробениуса матрицы: 32.07027779033304
Собственные числа матрицы: [31.83589099 -0.92134048  0.83803629]
Обратная матрица:
 [[-0.27422616  0.34395205 -0.06575735]
 [ 1.91340864 -0.74223505 -1.17254942]
 [-1.43674176  0.36850745  1.15576298]]


**5. (0.2 балла)** Повторите 100 раз следующий эксперимент: сгенерируйте две матрицы размера 10×10 из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль?

In [14]:
max_elements = []
for exp_num in range(100):
    matrix1 = np.random.normal(size=(10, 10))
    matrix2 = np.random.normal(size=(10, 10))

    product = np.dot(matrix1, matrix2)
    max_element = np.max(product)
    max_elements.append(max_element)

mean_max = np.mean(max_elements)
quantile_95 = np.percentile(max_elements, 95)
print("Среднее значение максимальных элементов:", mean_max)
print("95-процентная квантиль максимальных элементов:", quantile_95)

Среднее значение максимальных элементов: 8.107614242752994
95-процентная квантиль максимальных элементов: 10.958525054172199


### Pandas

![](https://metrouk2.files.wordpress.com/2015/10/panda.jpg)

#### Ответьте на вопросы о данных по авиарейсам в США за январь-апрель 2008 года.

Данные находятся в приложенном файле `2008.csv`. Их [описание](http://stat-computing.org/dataexpo/2009/the-data.html) приведено ниже:

Airline on-time performance

Have you ever been stuck in an airport because your flight was delayed or cancelled and wondered if you could have predicted it if you'd had more data? This is your chance to find out.

The data
The data set is available for download here.
The data consists of flight arrival and departure details for all commercial flights within the USA, from October 1987 to April 2008. This is a large dataset: there are nearly 120 million records in total, and takes up 1.6 gigabytes of space compressed and 12 gigabytes when uncompressed.

Understanding and preparing the data
In order to answer above questions, we are going to analyze the provided dataset, containing up to 1936758 ### different internal flights in the US for 2008 and their causes for delay, diversion and cancellation

The data comes from the U.S. Department of Transportation’s (DOT) Bureau of Transportation Statistics (BTS). Meta data explanations

This dataset is composed by the following variables:

**Year** 2008 **Month** 1-12 **DayofMonth** 1-31 **DayOfWeek** 1 (Monday) - 7 (Sunday)  
**DepTime** actual departure time (local, hhmm)  
**CRSDepTime** scheduled departure time (local, hhmm)  
**ArrTime** actual arrival time (local, hhmm)  
**CRSArrTime** scheduled arrival time (local, hhmm)  
**UniqueCarrier** unique carrier code  
**FlightNum** flight number  
**TailNum** plane tail number: aircraft registration, unique aircraft identifier  
**ActualElapsedTime** in minutes  
**CRSElapsedTime** in minutes  
**AirTime** in minutes  
**ArrDelay** arrival delay, in minutes: A flight is counted as “on time” if it operated less than 15 minutes later the scheduled time shown in the carriers’ Computerized Reservations Systems (CRS).  
**DepDelay** departure delay, in minutes  
**Origin** origin IATA airport code  
**Dest** destination IATA airport code  
**Distance** in miles  
**TaxiIn** taxi in time, in minutes  
**TaxiOut** taxi out time in minutes  
**Cancelled** *was the flight cancelled  
**CancellationCode** reason for cancellation (A = carrier, B = weather, C = NAS, D = security)  
**Diverted** 1 = yes, 0 = no  
**CarrierDelay** in minutes: Carrier delay is within the control of the air carrier. Examples of occurrences that may determine carrier delay are: aircraft cleaning, aircraft damage, awaiting the arrival of connecting passengers or crew, baggage, bird strike, cargo loading, catering, computer, outage-carrier equipment, crew legality (pilot or attendant rest), damage by hazardous goods, engineering inspection, fueling, handling disabled passengers, late crew, lavatory servicing, maintenance, oversales, potable water servicing, removal of unruly passenger, slow boarding or seating, stowing carry-on baggage, weight and balance delays.  
**WeatherDelay** in minutes: Weather delay is caused by extreme or hazardous weather conditions that are forecasted or manifest themselves on point of departure, enroute, or on point of arrival.  
**NASDelay** in minutes: Delay that is within the control of the National Airspace System (NAS) may include: non-extreme weather conditions, airport operations, heavy traffic volume, air traffic control, etc.  
**SecurityDelay** in minutes: Security delay is caused by evacuation of a terminal or concourse, re-boarding of aircraft because of security breach, inoperative screening equipment and/or long lines in excess of 29 minutes at screening areas.  
**LateAircraftDelay** in minutes: Arrival delay at an airport due to the late arrival of the same aircraft at a previous airport. The ripple effect of an earlier delay at downstream airports is referred to as delay propagation.

In [2]:
import pandas as pd
%matplotlib inline

**6. (0.3 балла)** Какая из причин отмены рейса (`CancellationCode`) была самой частой? (расшифровки кодов можно найти в описании данных)

In [6]:
data = pd.read_csv('2008.csv')

cancellation_counts = data['CancellationCode'].value_counts()
print('Самая частая причина отмены рейса', cancellation_counts.head(1), '\n', 'A = carrier (перевозчик) \n B = weather (погода) \n C = NAS (National Airspace System, национальная система воздушного пространства) \n D = security (безопасность)')

Самая частая причина отмены рейса CancellationCode
A    563
Name: count, dtype: int64 
 A = carrier (перевозчик) 
 B = weather (погода) 
 C = NAS (National Airspace System, национальная система воздушного пространства) 
 D = security (безопасность)


**7. (0.3 балла)** Найдите среднее, минимальное и максимальное расстояние, пройденное самолетом.

In [7]:
mean_distance = data['Distance'].mean()
min_distance = data['Distance'].min()
max_distance = data['Distance'].max()

print('Mean distance:', mean_distance)
print('Minimum distance:', min_distance)
print('Maximum distance:', max_distance)

Mean distance: 724.5082571428571
Minimum distance: 31
Maximum distance: 4962


**8. (0.3 балла)** Не выглядит ли подозрительным минимальное пройденное расстояние? В какие дни и на каких рейсах оно было? Какое расстояние было пройдено этими же рейсами в другие дни?

In [13]:
min_distance_by_flight = data.groupby('FlightNum')['Distance'].min()
min_distance = min_distance_by_flight.min()
min_distance_flights = data[data['Distance'] == min_distance]
print(data[data['Distance'] == min_distance][['FlightNum', 'Year', 'Month', 'DayofMonth', 'Distance']])

       FlightNum  Year  Month  DayofMonth  Distance
1116          65  2008     12          30        31
6958          65  2008     12          26        31
17349         64  2008      8          18        31
27534         64  2008      3          11        31
46082         65  2008      8           9        31
48112         64  2008      2          28        31


In [14]:
other_days = data.merge(min_distance_flights[['FlightNum']], on='FlightNum', how='inner')
other_days = other_days[other_days['Distance'] != min_distance]
print(other_days[['FlightNum', 'Year', 'Month', 'DayofMonth', 'Distance']])

     FlightNum  Year  Month  DayofMonth  Distance
0           64  2008      3          20       533
1           64  2008      3          20       533
2           64  2008      3          20       533
3           64  2008      3          13       680
4           64  2008      3          13       680
..         ...   ...    ...         ...       ...
229         65  2008      8           3      2454
230         65  2008      8           3      2454
231         65  2008      1           5      1005
232         65  2008      1           5      1005
233         65  2008      1           5      1005

[216 rows x 5 columns]


**9. (0.3 балла)** Из какого аэропорта было произведено больше всего вылетов? В каком городе он находится?

In [32]:
flights_by_origin = data.groupby('Origin').size().reset_index(name='NumFlights')
print(flights_by_origin.sort_values(by='NumFlights').tail(1))

   Origin  NumFlights
18    ATL        4134


**10. (0.3 балла)** Найдите для каждого аэропорта среднее время полета (`AirTime`) по всем вылетевшим из него рейсам. Какой аэропорт имеет наибольшее значение этого показателя?

In [47]:
mean_time = data.groupby("Origin")["AirTime"].mean()
print(mean_time)
print(mean_time.reset_index(name = 'MeanTime').sort_values(by = "MeanTime", ascending = False).head(1))

Origin
ABE    88.266667
ABI    36.400000
ABQ    93.454321
ABY    35.714286
ACK    50.800000
         ...    
WRG    18.000000
XNA    85.945736
YAK    35.900000
YKM    79.000000
YUM    47.470588
Name: AirTime, Length: 297, dtype: float64
    Origin  MeanTime
262    SJU     205.2


**11. (0.5 балла)** Найдите аэропорт, у которого наибольшая доля задержанных (`DepDelay > 0`) рейсов. Исключите при этом из рассмотрения аэропорты, из которых было отправлено меньше 1000 рейсов (используйте функцию `filter` после `groupby`).

In [44]:
flights_by_origin = data.groupby('Origin').size().reset_index(name='NumFlights')
flights_by_origin_names = list(flights_by_origin[flights_by_origin["NumFlights"] >= 1000]["Origin"])
big_flights_data = data[(data["Origin"].isin(flights_by_origin_names))]
grouped_flights_num = big_flights_data.groupby("Origin").size().reset_index(name = 'NumFlights')
grouped_flights_delay_num = big_flights_data[big_flights_data["DepDelay"] > 0].groupby("Origin").size().reset_index(name = 'NumFlights')
grouped_flights_delay_num['Ratio'] = grouped_flights_delay_num['NumFlights'] / grouped_flights_num['NumFlights']
grouped_flights_delay_num.sort_values(by = "Ratio").tail(1)

Origin  NumFlights     Ratio
7    EWR         710  0.511159

## Линейная регрессия

В этой части мы разберемся с линейной регрессией, способами её обучения и измерением качества ее прогнозов. 

Будем рассматривать датасет из предыдущей части задания для предсказания времени задержки отправления рейса в минутах (DepDelay). Отметим, что под задержкой подразумевается не только опоздание рейса относительно планируемого времени вылета, но и отправление до планируемого времени.

### Подготовка данных

**12. (0.5 балла)** Считайте выборку из файла при помощи функции pd.read_csv и ответьте на следующие вопросы:
   - Имеются ли в данных пропущенные значения?
   - Сколько всего пропущенных элементов в таблице "объект-признак"?
   - Сколько объектов имеют хотя бы один пропуск?
   - Сколько признаков имеют хотя бы одно пропущенное значение?

In [49]:
print('Количество пропусков - ', data.isnull().sum().sum())

Количество пропусков -  355215


In [55]:
one_null_data = data.isnull().sum(axis = 1).reset_index(name = "NanQ")
print('Количество объектов, которые имеют хотя бы один пропуск - ', one_null_data[one_null_data["NanQ"] != 0].shape[0])

       index  NanQ
0          0     6
1          1     6
2          2     6
3          3     6
4          4     6
...      ...   ...
69995  69995     6
69996  69996     1
69997  69997     6
69998  69998     1
69999  69999     6

[70000 rows x 2 columns]
Количество объектов, которые имеют хотя бы один пропуск -  70000


In [53]:
one_null_data = data.isnull().sum(axis = 0).reset_index(name = "NanQ")
print('Количество признаков, которые имеют хотя бы один пропуск - ', one_null_data[one_null_data["NanQ"] != 0].shape[0])

Количество признаков, которые имеют хотя бы один пропуск -  16


Как вы понимаете, также не имеет смысла рассматривать при решении поставленной задачи объекты с пропущенным значением целевой переменной. В связи с этим ответьте на следующие вопросы и выполните соответствующие действия:
- Имеются ли пропущенные значения в целевой переменной?
- Проанализируйте объекты с пропущенными значениями целевой переменной. Чем вызвано это явление? Что их объединяет? Можно ли в связи с этим, на ваш взгляд, исключить какие-то признаки из рассмотрения? Обоснуйте свою точку зрения.

Исключите из выборки объекты **с пропущенным значением целевой переменной и со значением целевой переменной, равным 0**, а также при необходимости исключите признаки в соответствии с вашим ответом на последний вопрос из списка и выделите целевую переменную в отдельный вектор, исключив её из матрицы "объект-признак".

In [59]:
if data["DepDelay"].isnull().any():
    print("Есть пропуски в целевой переменной")
else:
    print("Нет пропусков в целевой переменной")
    
cel = data[data["DepDelay"].isnull() == True]
pd.unique(cel["Cancelled"])

Есть пропуски в целевой переменной
11       1
78       1
103      1
112      1
128      1
        ..
69818    1
69836    1
69884    1
69918    1
69949    1
Name: Cancelled, Length: 1399, dtype: int64


array([1], dtype=int64)

In [60]:
index_list = list(data[(data["DepDelay"].isnull() == True) | (data["DepDelay"] == 0)].index)
data_linreg = data.drop(index_list, axis = 0)
data_linreg.drop(["CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay", "LateAircraftDelay", "Cancelled", 
                "CancellationCode", "Year", "TailNum"], axis = 1, inplace = True)
data_depdelay = data_linreg["DepDelay"]
data_linreg.drop("DepDelay", axis = 1, inplace = True)
data_linreg

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
0          6          18          3   2111.0        2055   2308.0        2300   
1          6           4          3   1426.0        1410   1730.0        1728   
2          8           3          7   1143.0        1145   1501.0        1520   
3          1          23          3   1141.0        1144   1323.0        1335   
4          5           4          7    815.0         820   1243.0        1300   
...      ...         ...        ...      ...         ...      ...         ...   
69994      1          26          6    920.0         925   1110.0        1118   
69995      5          12          1   1829.0        1840   2137.0        2149   
69996      5          11          7   1149.0        1125   1336.0        1314   
69998      2          18          1   1906.0        1900   2200.0        2139   
69999     12           6          6    859.0         900    959.0        1005   

      UniqueCarrier  FlightNum  ActualElapsedTime  CRSElapsedTime  AirTime  \
0                WN        242               57.0            65.0     46.0   
1                XE       2380              124.0           138.0    108.0   
2                WN       1769              138.0           155.0    125.0   
3                OO       3802              102.0           111.0     79.0   
4                WN        399              148.0           160.0    136.0   
...             ...        ...                ...             ...      ...   
69994            OH       5218              110.0           113.0     85.0   
69995            DL        794              128.0           129.0     86.0   
69996            OO       6159               47.0            49.0     30.0   
69998            NW        641              234.0           219.0    192.0   
69999            WN        510               60.0            65.0     46.0   

       ArrDelay Origin Dest  Distance  TaxiIn  TaxiOut  Diverted  
0           8.0    MDW  CLE       307     3.0      8.0         0  
1           2.0    IAH  AVL       834     4.0     12.0         0  
2         -19.0    MDW  TPA       997     4.0      9.0         0  
3         -12.0    SLC  FCA       532     4.0     19.0         0  
4         -17.0    LAS  AUS      1090     4.0      8.0         0  
...         ...    ...  ...       ...     ...      ...       ...  
69994      -8.0    CVG  BDL       661    12.0     13.0         0  
69995     -12.0    ORD  ATL       606    22.0     20.0         0  
69996      22.0    ORD  GRR       137     4.0     13.0         0  
69998      21.0    RSW  MSP      1416    20.0     22.0         0  
69999      -6.0    LAS  BUR       223     2.0     12.0         0  

[63404 rows x 19 columns]

**13. (0.5 балла)** Обратите внимание, что признаки DepTime, CRSDepTime, ArrTime, CRSArrTime приведены в формате hhmm, в связи с чем будет не вполне корректно рассматривать их как вещественные.

Преобразуйте каждый признак FeatureName из указанных в пару новых признаков FeatureName\_Hour, FeatureName\_Minute, разделив каждое из значений на часы и минуты. Не забудьте при этом исключить исходный признак из выборки. В случае, если значение признака отсутствует, значения двух новых признаков, его заменяющих, также должны отсутствовать. 

Например, признак DepTime необходимо заменить на пару признаков DepTime_Hour, DepTime_Minute. При этом, например, значение 155 исходного признака будет преобразовано в значения 1 и 55 признаков DepTime_Hour, DepTime_Minute соответственно.

Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  \
0          6          18          3            WN        242   
1          6           4          3            XE       2380   
2          8           3          7            WN       1769   
3          1          23          3            OO       3802   
4          5           4          7            WN        399   
...      ...         ...        ...           ...        ...   
69994      1          26          6            OH       5218   
69995      5          12          1            DL        794   
69996      5          11          7            OO       6159   
69998      2          18          1            NW        641   
69999     12           6          6            WN        510   

       ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay Origin  ...  \
0                   57.0            65.0     46.0       8.0    MDW  ...   
1                  124.0           138.0    108.0       2.0    IAH  ...   
2                  138.0           155.0    125.0     -19.0    MDW  ...   
3                  102.0           111.0     79.0     -12.0    SLC  ...   
4                  148.0           160.0    136.0     -17.0    LAS  ...   
...                  ...             ...      ...       ...    ...  ...   
69994              110.0           113.0     85.0      -8.0    CVG  ...   
69995              128.0           129.0     86.0     -12.0    ORD  ...   
69996               47.0            49.0     30.0      22.0    ORD  ...   
69998              234.0           219.0    192.0      21.0    RSW  ...   
69999               60.0            65.0     46.0      -6.0    LAS  ...   

      TaxiOut  Diverted  DepTime_Hour  DepTime_Minute  CRSDepTime_Hour  \
0         8.0         0          21.0            11.0             20.0   
1        12.0         0          14.0            26.0             14.0   
2         9.0         0          11.0            43.0             11.0   
3        19.0         0          11.0            41.0             11.0   
4         8.0         0           8.0            15.0              8.0   
...       ...       ...           ...             ...              ...   
69994    13.0         0           9.0            20.0              9.0   
69995    20.0         0          18.0            29.0             18.0   
69996    13.0         0          11.0            49.0             11.0   
69998    22.0         0          19.0             6.0             19.0   
69999    12.0         0           8.0            59.0              9.0   

       CRSDepTime_Minute  ArrTime_Hour  ArrTime_Minute  CRSArrTime_Hour  \
0                   55.0          23.0             8.0             23.0   
1                   10.0          17.0            30.0             17.0   
2                   45.0          15.0             1.0             15.0   
3                   44.0          13.0            23.0             13.0   
4                   20.0          12.0            43.0             13.0   
...                  ...           ...             ...              ...   
69994               25.0          11.0            10.0             11.0   
69995               40.0          21.0            37.0             21.0   
69996               25.0          13.0            36.0             13.0   
69998                0.0          22.0             0.0             21.0   
69999                0.0           9.0            59.0             10.0   

       CRSArrTime_Minute  
0                    0.0  
1                   28.0  
2                   20.0  
3                   35.0  
4                    0.0  
...                  ...  
69994               18.0  
69995               49.0  
69996               14.0  
69998               39.0  
69999                5.0  

[63404 rows x 23 columns]

**14. (0.5 балла)** Некоторые из признаков, отличных от целевой переменной, могут оказывать чересчур значимое влияние на прогноз, поскольку по своему смыслу содержат большую долю информации о значении целевой переменной. Изучите описание датасета и исключите признаки, сильно коррелирующие с ответами. Ваш выбор признаков для исключения из выборки обоснуйте. Кроме того, исключите признаки TailNum и Year.

In [ ]:
# Your code here

**15. (1 балл)** Приведем данные к виду, пригодному для обучения линейных моделей. Для этого вещественные признаки надо отмасштабировать, а категориальные — привести к числовому виду. Также надо устранить пропуски в данных.

В первую очередь поймем, зачем необходимо применять масштабирование. Следующие ячейки с кодом построят гистограммы для 3 вещественных признаков выборки.

In [ ]:
X['DepTime_Hour'].hist(bins=20)

In [ ]:
X['TaxiIn'].hist(bins=20)

In [ ]:
X['FlightNum'].hist(bins=20)

Какую проблему вы наблюдаете на этих графиках? Как масштабирование поможет её исправить?

Некоторые из признаков в нашем датасете являются категориальными. Типичным подходом к работе с ними является бинарное, или [one-hot-кодирование](https://en.wikipedia.org/wiki/One-hot).

Реализуйте функцию transform_data, которая принимает на вход DataFrame с признаками и выполняет следующие шаги:
1. Замена пропущенных значений на нули для вещественных признаков и на строки 'nan' для категориальных.
2. Масштабирование вещественных признаков с помощью [StandardScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html).
3. One-hot-кодирование категориальных признаков с помощью [DictVectorizer](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.DictVectorizer.html) или функции [pd.get_dummies](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html).

Метод должен возвращать преобразованный DataFrame, который должна состоять из масштабированных вещественных признаков и закодированных категориальных (исходные признаки должны быть исключены из выборки).

In [ ]:
def transform_data(data):
    # Your code here

Примените функцию transform_data к данным. Сколько признаков получилось после преобразования?

In [ ]:
# Your code here

**16. (0.5 балла)** Разбейте выборку и вектор целевой переменной на обучение и контроль в отношении 70/30 (для этого можно использовать, например, функцию [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.train_test_split.html)). 

In [ ]:
# Your code here

### Scikit-learn

<img src = "https://pp.vk.me/c4534/u35727827/93547647/x_d31c4463.jpg">
Теперь, когда мы привели данные к пригодному виду, попробуем решить задачу при помощи метода наименьших квадратов. Напомним, что данный метод заключается в оптимизации функционала $MSE$:

$$MSE(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 \to \min_{w},$$

где $\{ (x_i, y_i ) \}_{i=1}^l$ — обучающая выборка, состоящая из $l$ пар объект-ответ.

Заметим, что решение данной задачи уже реализовано в модуле sklearn в виде класса [LinearRegression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression).

**17. (0.5 балла)** Обучите линейную регрессию на 1000 объектах из обучающей выборки и выведите значения $MSE$ и $R^2$ на этой подвыборке и контрольной выборке (итого 4 различных числа). Проинтерпретируйте полученный результат — насколько качественные прогнозы строит полученная модель? Какие проблемы наблюдаются в модели?

**Подсказка**: изучите значения полученных коэффициентов $w$, сохраненных в атрибуте coef_ объекта LinearRegression.

In [ ]:
# Your code here

Для решения описанных вами в предыдущем пункте проблем используем L1- или L2-регуляризацию, тем самым получив Lasso и Ridge регрессии соответственно и изменив оптимизационную задачу одним из следующих образов:
$$MSE_{L1}(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 + \alpha ||w||_1 \to \min_{w},$$
$$MSE_{L2}(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 + \alpha ||w||_2^2 \to \min_{w},$$

где $\alpha$ — коэффициент регуляризации. Один из способов его подбора заключается в переборе некоторого количества значений и оценке качества на кросс-валидации для каждого из них, после чего выбирается значение, для которого было получено наилучшее качество.

In [ ]:
# Your code here

__18. (1 балл) __ Обучение линейной регрессии.



Обучите линейную регрессию с $L_1$ (Lasso) и $L_2$ (Ridge) регуляризаторами (используйте параметры по умолчанию). Посмотрите, какое количество коэффициентов близко к 0 (степень близости к 0 определите сами из разумных пределов). Постройте график зависимости числа ненулевых коэффициентов от коэффицента регуляризации (перебирайте значения по логарифмической сетке от $10^{-3}$ до $10^3$). Согласуются ли результаты с вашими ожиданиями?

In [ ]:
# Your code here
# ...

Посчитайте для Ridge-регрессии следующие метрики: $RMSE$, $MAE$, $R^2$.

In [ ]:
# Your code here
# ...

Подберите на обучающей выборке для Ridge-регрессии коэффициент регуляризации (перебирайте значения по логарифмической сетке от $10^{-3}$ до $10^3$) для каждой из метрик при помощи кросс-валидации c 5 фолдами на тех же 1000 объектах. Для этого воспользуйтесь GridSearchCV и KFold из sklearn. Постройте графики зависимости фукнции потерь от коэффициента регуляризации. Посчитайте те же метрики снова. Заметно ли изменилось качество?

Для выполнения данного задания вам могут понадобиться реализованные в библиотеке объекты [LassoCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html), [RidgeCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html) и [KFold](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.KFold.html).

In [ ]:
# Your code here
# ...

__19. (0.5 балла)__ Поиск объектов-выбросов


Как известно, MSE сильно штрафует за большие ошибки на объектах-выбросах. С помощью cross_val_predict сделайте Out-of-Fold предсказания для обучающей выборки. Посчитайте ошибки и посмотрите на их распределение (plt.hist). Что вы видите?